src/mscorlib/shared/System/Threading/AsyncLocal.cs

```diff
@@ -124,14 +124,29 @@ internal static class AsyncLocalValueMap
     {
         public static IAsyncLocalValueMap Empty { get; } = new EmptyAsyncLocalValueMap();
 
+        public static bool IsEmpty(IAsyncLocalValueMap asyncLocalValueMap)
+        {
+            Debug.Assert(asyncLocalValueMap != null);
+            return asyncLocalValueMap == Empty;
+        }
+
+        public static IAsyncLocalValueMap Create(IAsyncLocal key, object value, bool treatNullValueAsNonexistent)
+        {
+            // If the value isn't null or a null value may not be treated as nonexistent, then create a new one-element map
+            // to store the key/value pair.  Otherwise, use the empty map.
+            return !(value == null && treatNullValueAsNonexistent) ?
```

Comment:
FWIW, I'd personally find the condition easier to understand as:
```
value != null || !treatNullValueAsNonexistent
```
Same in other cases.  Up to you, though.